In [2]:
import pandas as pd

Read data from clean data csv.

In [3]:
file = r'data/clean_data.csv'
df = pd.read_csv(file)
df

date  mintemp  maxtemp   meantemp  minpres  maxpres    meanpres  \
0     27.09.2012     12.0     21.0  15.958333    761.2    763.5  762.487500   
1     28.09.2012     15.0     24.0  19.000000    762.0    765.0  763.616667   
2     29.09.2012      8.0     25.0  16.875000    761.2    765.0  763.541667   
3     30.09.2012      8.0     26.0  16.833333    760.5    765.0  763.037500   
4      1.10.2012      1.0     24.0  11.125000    759.7    768.1  763.915909   
...          ...      ...      ...        ...      ...      ...         ...   
3712  26.11.2022     -5.0      4.0   0.333333    763.5    769.6  767.350000   
3713  27.11.2022     -2.0      8.0   3.437500    759.0    763.5  761.060417   
3714  28.11.2022     -2.0     13.0   6.479167    760.5    772.7  763.922917   
3715  29.11.2022    -10.0     -2.0  -5.958333    772.7    783.1  778.268750   
3716  30.11.2022    -11.0     -2.0  -6.229167    777.7    783.1  780.079167   

      maxhumi  minhumi   meanhumi  ...  S_wd  SE_wd  E_wd  NE_wd  nocld  \
0       100.0     39.0  75.375000  ...     0      0     0      0      1   
1        42.0     10.0  22.875000  ...     0      0     0      0      1   
2        82.0     17.0  43.583333  ...     1      1     1      1      1   
3        87.0     18.0  52.750000  ...     0      0     0      0      1   
4       100.0     11.0  64.204545  ...     1      1     1      1      1   
...       ...      ...        ...  ...   ...    ...   ...    ...    ...   
3712     50.0     14.0  26.770833  ...     0      0     0      0      1   
3713     56.0     23.0  36.208333  ...     0      0     0      0      1   
3714     61.0     17.0  31.729167  ...     0      0     0      0      1   
3715     30.0     12.0  21.208333  ...     0      0     0      0      1   
3716     57.0     14.0  23.166667  ...     0      0     0      0      1   

      little  some  much  cloudy  rain  
0          0     0     0       0     0  
1          0     0     0       0     0  
2          0     0     0       0     0  
3          0     0     0       0     0  
4          0     0     0       0     0  
...      ...   ...   ...     ...   ...  
3712       0     0     0       0     0  
3713       0     0     0       0     0  
3714       0     0     0       0     0  
3715       0     0     0       0     0  
3716       0     0     0       0     0  

[3717 rows x 28 columns]

Create small sample DataFrame for easier testing.

In [4]:
tmp = df[['date', 'mintemp', 'meanpres', 'E_wd', 'some']].head(10)
tmp

date  mintemp    meanpres  E_wd  some
0  27.09.2012     12.0  762.487500     0     0
1  28.09.2012     15.0  763.616667     0     0
2  29.09.2012      8.0  763.541667     1     0
3  30.09.2012      8.0  763.037500     0     0
4   1.10.2012      1.0  763.915909     1     0
5   2.10.2012      6.0  761.095833     1     0
6   3.10.2012      2.0  761.276389     0     0
7   4.10.2012      3.0  762.133333     1     0
8   5.10.2012      6.0  764.247945     1     0
9   6.10.2012      7.0  765.305556     1     0

In [5]:
# 1 day prior
N = 2

# target measurement of mean temperature
feature = 'mintemp'

# total number of rows
rows = tmp.shape[0]

# a list representing Nth prior measurements of feature
# notice that the front of the list needs to be padded with N
# None values to maintain the constistent rows length for each N
nth_prior_measurements = [None]*N + [tmp[feature][i-N] for i in range(N, rows)]

# make a new column name of feature_N and add to DataFrame
col_name = "{}_{}".format(feature, N)  
tmp[col_name] = nth_prior_measurements  
tmp  

date  mintemp    meanpres  E_wd  some  mintemp_2
0  27.09.2012     12.0  762.487500     0     0        NaN
1  28.09.2012     15.0  763.616667     0     0        NaN
2  29.09.2012      8.0  763.541667     1     0       12.0
3  30.09.2012      8.0  763.037500     0     0       15.0
4   1.10.2012      1.0  763.915909     1     0        8.0
5   2.10.2012      6.0  761.095833     1     0        8.0
6   3.10.2012      2.0  761.276389     0     0        1.0
7   4.10.2012      3.0  762.133333     1     0        6.0
8   5.10.2012      6.0  764.247945     1     0        2.0
9   6.10.2012      7.0  765.305556     1     0        3.0

Function for creating the required feature for the n-th prior day.

In [6]:
def get_nth_prior(df, feature, n): 
    rows = df.shape[0]
    nth_prior_measurements = [None]*n + [df[feature][i-n] for i in range(n, rows)]
    col_name = "{}_{}".format(feature, n)
    df[col_name] = nth_prior_measurements

In [7]:
features = [column for column in tmp]
print(features)

for feature in features:
    if feature != 'date':
        for n in range(1, 4):
            get_nth_prior(tmp, feature, n)

tmp

['date', 'mintemp', 'meanpres', 'E_wd', 'some', 'mintemp_2']


date  mintemp    meanpres  E_wd  some  mintemp_2  mintemp_1  \
0  27.09.2012     12.0  762.487500     0     0        NaN        NaN   
1  28.09.2012     15.0  763.616667     0     0        NaN       12.0   
2  29.09.2012      8.0  763.541667     1     0       12.0       15.0   
3  30.09.2012      8.0  763.037500     0     0       15.0        8.0   
4   1.10.2012      1.0  763.915909     1     0        8.0        8.0   
5   2.10.2012      6.0  761.095833     1     0        8.0        1.0   
6   3.10.2012      2.0  761.276389     0     0        1.0        6.0   
7   4.10.2012      3.0  762.133333     1     0        6.0        2.0   
8   5.10.2012      6.0  764.247945     1     0        2.0        3.0   
9   6.10.2012      7.0  765.305556     1     0        3.0        6.0   

   mintemp_3  meanpres_1  meanpres_2  meanpres_3  E_wd_1  E_wd_2  E_wd_3  \
0        NaN         NaN         NaN         NaN     NaN     NaN     NaN   
1        NaN  762.487500         NaN         NaN     0.0     NaN     NaN   
2        NaN  763.616667  762.487500         NaN     0.0     0.0     NaN   
3       12.0  763.541667  763.616667  762.487500     1.0     0.0     0.0   
4       15.0  763.037500  763.541667  763.616667     0.0     1.0     0.0   
5        8.0  763.915909  763.037500  763.541667     1.0     0.0     1.0   
6        8.0  761.095833  763.915909  763.037500     1.0     1.0     0.0   
7        1.0  761.276389  761.095833  763.915909     0.0     1.0     1.0   
8        6.0  762.133333  761.276389  761.095833     1.0     0.0     1.0   
9        2.0  764.247945  762.133333  761.276389     1.0     1.0     0.0   

   some_1  some_2  some_3  mintemp_2_1  mintemp_2_2  mintemp_2_3  
0     NaN     NaN     NaN          NaN          NaN          NaN  
1     0.0     NaN     NaN          NaN          NaN          NaN  
2     0.0     0.0     NaN          NaN          NaN          NaN  
3     0.0     0.0     0.0         12.0          NaN          NaN  
4     0.0     0.0     0.0         15.0         12.0          NaN  
5     0.0     0.0     0.0          8.0         15.0         12.0  
6     0.0     0.0     0.0          8.0          8.0         15.0  
7     0.0     0.0     0.0          1.0          8.0          8.0  
8     0.0     0.0     0.0          6.0          1.0          8.0  
9     0.0     0.0     0.0          2.0          6.0          1.0

In [8]:
features = [column for column in df]
print(features)

for feature in features:
    if feature != 'date':
        for n in range(1, 4):
            get_nth_prior(df, feature, n)

df

['date', 'mintemp', 'maxtemp', 'meantemp', 'minpres', 'maxpres', 'meanpres', 'maxhumi', 'minhumi', 'meanhumi', 'meanwdspd', 'maxdewpt', 'mindewpt', 'meandewpt', 'N_wd', 'NW_wd', 'W_wd', 'SW_wd', 'S_wd', 'SE_wd', 'E_wd', 'NE_wd', 'nocld', 'little', 'some', 'much', 'cloudy', 'rain']


date  mintemp  maxtemp   meantemp  minpres  maxpres    meanpres  \
0     27.09.2012     12.0     21.0  15.958333    761.2    763.5  762.487500   
1     28.09.2012     15.0     24.0  19.000000    762.0    765.0  763.616667   
2     29.09.2012      8.0     25.0  16.875000    761.2    765.0  763.541667   
3     30.09.2012      8.0     26.0  16.833333    760.5    765.0  763.037500   
4      1.10.2012      1.0     24.0  11.125000    759.7    768.1  763.915909   
...          ...      ...      ...        ...      ...      ...         ...   
3712  26.11.2022     -5.0      4.0   0.333333    763.5    769.6  767.350000   
3713  27.11.2022     -2.0      8.0   3.437500    759.0    763.5  761.060417   
3714  28.11.2022     -2.0     13.0   6.479167    760.5    772.7  763.922917   
3715  29.11.2022    -10.0     -2.0  -5.958333    772.7    783.1  778.268750   
3716  30.11.2022    -11.0     -2.0  -6.229167    777.7    783.1  780.079167   

      maxhumi  minhumi   meanhumi  ...  some_3  much_1  much_2  much_3  \
0       100.0     39.0  75.375000  ...     NaN     NaN     NaN     NaN   
1        42.0     10.0  22.875000  ...     NaN     0.0     NaN     NaN   
2        82.0     17.0  43.583333  ...     NaN     0.0     0.0     NaN   
3        87.0     18.0  52.750000  ...     0.0     0.0     0.0     0.0   
4       100.0     11.0  64.204545  ...     0.0     0.0     0.0     0.0   
...       ...      ...        ...  ...     ...     ...     ...     ...   
3712     50.0     14.0  26.770833  ...     0.0     0.0     0.0     0.0   
3713     56.0     23.0  36.208333  ...     0.0     0.0     0.0     0.0   
3714     61.0     17.0  31.729167  ...     0.0     0.0     0.0     0.0   
3715     30.0     12.0  21.208333  ...     0.0     0.0     0.0     0.0   
3716     57.0     14.0  23.166667  ...     0.0     0.0     0.0     0.0   

      cloudy_1  cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
0          NaN       NaN       NaN     NaN     NaN     NaN  
1          0.0       NaN       NaN     0.0     NaN     NaN  
2          0.0       0.0       NaN     0.0     0.0     NaN  
3          0.0       0.0       0.0     0.0     0.0     0.0  
4          0.0       0.0       0.0     0.0     0.0     0.0  
...        ...       ...       ...     ...     ...     ...  
3712       0.0       0.0       0.0     0.0     0.0     0.0  
3713       0.0       0.0       0.0     0.0     0.0     0.0  
3714       0.0       0.0       0.0     0.0     0.0     0.0  
3715       0.0       0.0       0.0     0.0     0.0     0.0  
3716       0.0       0.0       0.0     0.0     0.0     0.0  

[3717 rows x 109 columns]

In [9]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3717 entries, 0 to 3716
Data columns (total 109 columns):
 #    Column       Dtype  
---   ------       -----  
 0    date         object 
 1    mintemp      float64
 2    maxtemp      float64
 3    meantemp     float64
 4    minpres      float64
 5    maxpres      float64
 6    meanpres     float64
 7    maxhumi      float64
 8    minhumi      float64
 9    meanhumi     float64
 10   meanwdspd    float64
 11   maxdewpt     float64
 12   mindewpt     float64
 13   meandewpt    float64
 14   N_wd         int64  
 15   NW_wd        int64  
 16   W_wd         int64  
 17   SW_wd        int64  
 18   S_wd         int64  
 19   SE_wd        int64  
 20   E_wd         int64  
 21   NE_wd        int64  
 22   nocld        int64  
 23   little       int64  
 24   some         int64  
 25   much         int64  
 26   cloudy       int64  
 27   rain         int64  
 28   mintemp_1    float64
 29   mintemp_2    float64
 30   mintemp_3    float64
 

Drop rows with null value to ensure the integrity of dataset.

In [10]:
df_nona = df.dropna()
df_nona

date  mintemp  maxtemp   meantemp  minpres  maxpres    meanpres  \
3     30.09.2012      8.0     26.0  16.833333    760.5    765.0  763.037500   
4      1.10.2012      1.0     24.0  11.125000    759.7    768.1  763.915909   
5      2.10.2012      6.0     25.0  14.513889    758.2    765.0  761.095833   
6      3.10.2012      2.0     25.0  14.027778    758.2    765.8  761.276389   
7      4.10.2012      3.0     25.0  14.111111    756.9    768.1  762.133333   
...          ...      ...      ...        ...      ...      ...         ...   
3712  26.11.2022     -5.0      4.0   0.333333    763.5    769.6  767.350000   
3713  27.11.2022     -2.0      8.0   3.437500    759.0    763.5  761.060417   
3714  28.11.2022     -2.0     13.0   6.479167    760.5    772.7  763.922917   
3715  29.11.2022    -10.0     -2.0  -5.958333    772.7    783.1  778.268750   
3716  30.11.2022    -11.0     -2.0  -6.229167    777.7    783.1  780.079167   

      maxhumi  minhumi   meanhumi  ...  some_3  much_1  much_2  much_3  \
3        87.0     18.0  52.750000  ...     0.0     0.0     0.0     0.0   
4       100.0     11.0  64.204545  ...     0.0     0.0     0.0     0.0   
5       100.0     24.0  66.888889  ...     0.0     0.0     0.0     0.0   
6       100.0      8.0  55.013889  ...     0.0     0.0     0.0     0.0   
7       100.0     13.0  52.069444  ...     0.0     0.0     0.0     0.0   
...       ...      ...        ...  ...     ...     ...     ...     ...   
3712     50.0     14.0  26.770833  ...     0.0     0.0     0.0     0.0   
3713     56.0     23.0  36.208333  ...     0.0     0.0     0.0     0.0   
3714     61.0     17.0  31.729167  ...     0.0     0.0     0.0     0.0   
3715     30.0     12.0  21.208333  ...     0.0     0.0     0.0     0.0   
3716     57.0     14.0  23.166667  ...     0.0     0.0     0.0     0.0   

      cloudy_1  cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
3          0.0       0.0       0.0     0.0     0.0     0.0  
4          0.0       0.0       0.0     0.0     0.0     0.0  
5          0.0       0.0       0.0     0.0     0.0     0.0  
6          0.0       0.0       0.0     0.0     0.0     0.0  
7          0.0       0.0       0.0     0.0     0.0     0.0  
...        ...       ...       ...     ...     ...     ...  
3712       0.0       0.0       0.0     0.0     0.0     0.0  
3713       0.0       0.0       0.0     0.0     0.0     0.0  
3714       0.0       0.0       0.0     0.0     0.0     0.0  
3715       0.0       0.0       0.0     0.0     0.0     0.0  
3716       0.0       0.0       0.0     0.0     0.0     0.0  

[3714 rows x 109 columns]

In [11]:
import re

s = r'._\d'

to_keep = df_nona.columns[df_nona.columns.str.contains(s, regex=True)]
to_remove = [column for column in df_nona.columns if column not in to_keep]
x = df_nona.drop(to_remove, axis=1)
x

mintemp_1  mintemp_2  mintemp_3  maxtemp_1  maxtemp_2  maxtemp_3  \
3           8.0       15.0       12.0       25.0       24.0       21.0   
4           8.0        8.0       15.0       26.0       25.0       24.0   
5           1.0        8.0        8.0       24.0       26.0       25.0   
6           6.0        1.0        8.0       25.0       24.0       26.0   
7           2.0        6.0        1.0       25.0       25.0       24.0   
...         ...        ...        ...        ...        ...        ...   
3712        1.0        1.0       -1.0       13.0        9.0       13.0   
3713       -5.0        1.0        1.0        4.0       13.0        9.0   
3714       -2.0       -5.0        1.0        8.0        4.0       13.0   
3715       -2.0       -2.0       -5.0       13.0        8.0        4.0   
3716      -10.0       -2.0       -2.0       -2.0       13.0        8.0   

      meantemp_1  meantemp_2  meantemp_3  minpres_1  ...  some_3  much_1  \
3      16.875000   19.000000   15.958333      761.2  ...     0.0     0.0   
4      16.833333   16.875000   19.000000      760.5  ...     0.0     0.0   
5      11.125000   16.833333   16.875000      759.7  ...     0.0     0.0   
6      14.513889   11.125000   16.833333      758.2  ...     0.0     0.0   
7      14.027778   14.513889   11.125000      758.2  ...     0.0     0.0   
...          ...         ...         ...        ...  ...     ...     ...   
3712    7.708333    6.625000    4.916667      759.0  ...     0.0     0.0   
3713    0.333333    7.708333    6.625000      763.5  ...     0.0     0.0   
3714    3.437500    0.333333    7.708333      759.0  ...     0.0     0.0   
3715    6.479167    3.437500    0.333333      760.5  ...     0.0     0.0   
3716   -5.958333    6.479167    3.437500      772.7  ...     0.0     0.0   

      much_2  much_3  cloudy_1  cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
3        0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
4        0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
5        0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
6        0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
7        0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
...      ...     ...       ...       ...       ...     ...     ...     ...  
3712     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3713     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3714     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3715     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3716     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  

[3714 rows x 81 columns]

In [12]:
y = df_nona['meantemp']
y

3       16.833333
4       11.125000
5       14.513889
6       14.027778
7       14.111111
          ...    
3712     0.333333
3713     3.437500
3714     6.479167
3715    -5.958333
3716    -6.229167
Name: meantemp, Length: 3714, dtype: float64

In [15]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)
x_train

mintemp_1  mintemp_2  mintemp_3  maxtemp_1  maxtemp_2  maxtemp_3  \
2501       18.0       16.0       22.0       31.0       34.0       37.0   
1969       -9.0      -10.0       -5.0        7.0        5.0        5.0   
599        13.0       14.0       14.0       30.0       28.0       30.0   
1697       20.0       21.0       17.0       35.0       37.0       36.0   
3403       -7.0       -8.0      -12.0       -2.0       -2.0        3.0   
...         ...        ...        ...        ...        ...        ...   
1133       -6.0       -4.0       -1.0       19.0       20.0       19.0   
1297        6.0       13.0        8.0       22.0       23.0       27.0   
863        -7.0       -6.0       -6.0       13.0        9.0       11.0   
3510        6.0        6.0        7.0       36.0       35.0       31.0   
3177       17.0       20.0       17.0       35.0       34.0       34.0   

      meantemp_1  meantemp_2  meantemp_3  minpres_1  ...  some_3  much_1  \
2501   25.055556   26.135417   29.583333      747.0  ...     0.0     0.0   
1969   -0.500000   -1.375000    0.437500      764.3  ...     0.0     0.0   
599    23.400000   22.222222   23.361702      753.1  ...     0.0     0.0   
1697   28.250000   29.083333   27.270833      754.6  ...     0.0     0.0   
3403   -4.166667   -3.583333   -4.395833      774.7  ...     0.0     0.0   
...          ...         ...         ...        ...  ...     ...     ...   
1133    4.491803    4.830645    5.049587      764.3  ...     0.0     0.0   
1297   16.060606   17.878788   16.562500      756.2  ...     0.0     0.0   
863     2.713115    2.754237    1.159664      762.0  ...     0.0     0.0   
3510   20.388889   19.506944   22.722222      746.3  ...     0.0     0.0   
3177   26.381944   26.243056   23.166667      747.8  ...     0.0     0.0   

      much_2  much_3  cloudy_1  cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
2501     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
1969     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
599      0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
1697     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3403     0.0     0.0       1.0       0.0       0.0     0.0     0.0     0.0  
...      ...     ...       ...       ...       ...     ...     ...     ...  
1133     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
1297     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
863      0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3510     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  
3177     0.0     0.0       0.0       0.0       0.0     0.0     0.0     0.0  

[2971 rows x 81 columns]

In [16]:
y_train

2501    25.694444
1969    -2.062500
599     22.468085
1697    24.333333
3403    -3.625000
          ...    
1133     3.905512
1297    15.588235
863      1.789474
3510    20.763889
3177    25.352113
Name: meantemp, Length: 2971, dtype: float64

In [17]:
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()

regressor.fit(x_train, y_train)

LinearRegression()

In [19]:
prediction = regressor.predict(x_test)

In [20]:
from sklearn.metrics import mean_absolute_error, median_absolute_error  
print("The Explained Variance: %.2f" % regressor.score(x_test, y_test))  
print("The Mean Absolute Error: %.2f degrees celsius" % mean_absolute_error(y_test, prediction))  
print("The Median Absolute Error: %.2f degrees celsius" % median_absolute_error(y_test, prediction))

The Explained Variance: 0.97
The Mean Absolute Error: 1.45 degrees celsius
The Median Absolute Error: 1.08 degrees celsius


### Testing zone

In [27]:
test_file = r'data/clean_data_test.csv'
df_test = pd.read_csv(test_file)
df_test = df_test.append({}, ignore_index=True)
df_test

/tmp/ipykernel_1977/554140614.py:3: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df_test = df_test.append({}, ignore_index=True)
/tmp/ipykernel_1977/554140614.py:3: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  df_test = df_test.append({}, ignore_index=True)


date  mintemp  maxtemp  meantemp  minpres  maxpres    meanpres  \
0  30.11.2022    -11.0     -2.0 -6.229167    777.7    783.1  780.079167   
1   1.12.2022    -14.0      1.0 -5.666667    773.9    778.5  775.935417   
2   2.12.2022    -13.0      0.0 -7.416667    766.6    773.4  769.662500   
3         NaN      NaN      NaN       NaN      NaN      NaN         NaN   

   maxhumi  minhumi   meanhumi  ...  S_wd  SE_wd  E_wd  NE_wd  nocld  little  \
0     57.0     14.0  23.166667  ...   0.0    0.0   0.0    0.0    1.0     0.0   
1     72.0     10.0  37.270833  ...   0.0    0.0   0.0    0.0    1.0     0.0   
2     79.0     37.0  59.187500  ...   1.0    1.0   1.0    1.0    1.0     0.0   
3      NaN      NaN        NaN  ...   NaN    NaN   NaN    NaN    NaN     NaN   

   some  much  cloudy  rain  
0   0.0   0.0     0.0   0.0  
1   0.0   0.0     0.0   0.0  
2   0.0   0.0     0.0   0.0  
3   NaN   NaN     NaN   NaN  

[4 rows x 28 columns]

In [28]:
features = [column for column in df_test]
print(features)

for feature in features:
    if feature != 'date':
        for n in range(1, 4):
            get_nth_prior(df_test, feature, n)

df_test

['date', 'mintemp', 'maxtemp', 'meantemp', 'minpres', 'maxpres', 'meanpres', 'maxhumi', 'minhumi', 'meanhumi', 'meanwdspd', 'maxdewpt', 'mindewpt', 'meandewpt', 'N_wd', 'NW_wd', 'W_wd', 'SW_wd', 'S_wd', 'SE_wd', 'E_wd', 'NE_wd', 'nocld', 'little', 'some', 'much', 'cloudy', 'rain']


date  mintemp  maxtemp  meantemp  minpres  maxpres    meanpres  \
0  30.11.2022    -11.0     -2.0 -6.229167    777.7    783.1  780.079167   
1   1.12.2022    -14.0      1.0 -5.666667    773.9    778.5  775.935417   
2   2.12.2022    -13.0      0.0 -7.416667    766.6    773.4  769.662500   
3         NaN      NaN      NaN       NaN      NaN      NaN         NaN   

   maxhumi  minhumi   meanhumi  ...  some_3  much_1  much_2  much_3  cloudy_1  \
0     57.0     14.0  23.166667  ...     NaN     NaN     NaN     NaN       NaN   
1     72.0     10.0  37.270833  ...     NaN     0.0     NaN     NaN       0.0   
2     79.0     37.0  59.187500  ...     NaN     0.0     0.0     NaN       0.0   
3      NaN      NaN        NaN  ...     0.0     0.0     0.0     0.0       0.0   

   cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
0       NaN       NaN     NaN     NaN     NaN  
1       NaN       NaN     0.0     NaN     NaN  
2       0.0       NaN     0.0     0.0     NaN  
3       0.0       0.0     0.0     0.0     0.0  

[4 rows x 109 columns]

In [32]:
df_test_nona = df_test.drop([0,1,2], axis=0).dropna(axis=1)
df_test_nona

mintemp_1  mintemp_2  mintemp_3  maxtemp_1  maxtemp_2  maxtemp_3  \
3      -13.0      -14.0      -11.0        0.0        1.0       -2.0   

   meantemp_1  meantemp_2  meantemp_3  minpres_1  ...  some_3  much_1  much_2  \
3   -7.416667   -5.666667   -6.229167      766.6  ...     0.0     0.0     0.0   

   much_3  cloudy_1  cloudy_2  cloudy_3  rain_1  rain_2  rain_3  
3     0.0       0.0       0.0       0.0     0.0     0.0     0.0  

[1 rows x 81 columns]

In [33]:
regressor.predict(df_test_nona)

array([-5.4708124])

20221203 offcial weather forecast:

![weather_forecast_20221203](bj20221203weather.png)

mean temperature: $\frac{2 + (-8)}{2} = -3$

predicted mean temperature: $-5.47$

error: $2.47$